opt: support column directions

RaduBerinde opened the pull request:
#### opt: allow specifying column directions in testcases

Switching the opt test syntax to use a
`column=(type [ascending|descending], ...)`
argument.

Release note: None

#### opt: support column directions

Adding support for column directions. This will need to be taken into
account in a bunch of cases so it makes sense to do it early.

At a high level, the column direction simply negates the result of a
comparison on that column.

Release note: None

RaduBerinde committed:
opt: allow specifying column directions in testcases

Switching the opt test syntax to use a
`column=(type [ascending|descending], ...)`
argument.

Release note: None

RaduBerinde committed:
opt: support column directions

Adding support for column directions. This will need to be taken into
account in a bunch of cases so it makes sense to do it early.

At a high level, the column direction simply negates the result of a
comparison on that column.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20897)
<!-- Reviewable:end -->

knz commented:
First commit LGTM with nits.

The second commit I don't understand. Why do you need to add this depth/offset parameter everywhere to take column directions into account?

---

Reviewed 4 of 4 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/index_constraints.go, line 159 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0krOhf_DmFq41aIyiJ:-L0krOhf_DmFq41aIyiK:b7lexu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b819d2ce176f290dcf2ba2b83e6dc68b33df59f4/pkg/sql/opt/index_constraints.go#L159)):*
> ```Go
> 
> // compareKeyVals compares two lists of Datums that map to the same
> // sequence of index columns (starting at depth).
> ```

Please explain "map to the same sequence of index columns" -- the method here simply compares the datums according to the ordering given in `c.colInfos`?

Also the word "depth" here and below is misleading - I believe you mean "offset". The things don't form a two dimensional data structure, so the notion of "depth" is not defined.

---

*[pkg/sql/opt/opt_test.go, line 150 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0kqRquGOqrXXqhpXI4:-L0kqRqvQJmyzG1daCdi:bx68tyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/517aacf7b55788039b83360fc3a615596b136d81/pkg/sql/opt/opt_test.go#L150)):*
> ```Go
> }
> 
> var splitDirectivesRE = regexp.MustCompile("^ *[a-zA-Z_,-]+(|=[a-zA-Z_]+|=([^_]*))( |$)")
> ```

This does not match what you think it matches. (At least it doesn't match what you describe below.)
Can you provide some more examples of what you intend to support? I can help build the re with you.

---

*[pkg/sql/opt/opt_test.go, line 180 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0kqj1M6I3YLsKGJJ_d:-L0kqj1M6I3YLsKGJJ_e:b-5xl4fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/517aacf7b55788039b83360fc3a615596b136d81/pkg/sql/opt/opt_test.go#L180)):*
> ```Go
> 		r.emit(line)
> 
> 		if strings.HasPrefix(line, "#") {
> ```

`strings.HasPrefix(strings.TrimLeft(line, " "), "#")`?
Make a function (`stripComments`).

---

*[pkg/sql/opt/opt_test.go, line 361 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0kr6NxRb7aiD7GKLTD:-L0kr6Nykig0jx2gRyZV:bn32cnc) ([raw file](https://github.com/cockroachdb/cockroach/blob/517aacf7b55788039b83360fc3a615596b136d81/pkg/sql/opt/opt_test.go#L361)):*
> ```Go
> }
> 
> func parseColumns(colStrs []string) ([]IndexColumnInfo, error) {
> ```

These various functions should be documented.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-:-L0krqwe_TdB07kbi1Fa:b8i14ip)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Because I need to know *which* column directions to take into account. There is a comment for LogicalSpan that explains that internally we use LogicalSpans for index columns starting at a given offset. Updated, hopefully it's more clear now.

---

Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/index_constraints.go, line 159 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0krOhf_DmFq41aIyiJ:-L0kuHg4Q-J2BO7Ddhrs:b-rg0jmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b819d2ce176f290dcf2ba2b83e6dc68b33df59f4/pkg/sql/opt/index_constraints.go#L159)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please explain "map to the same sequence of index columns" -- the method here simply compares the datums according to the ordering given in `c.colInfos`?

Also the word "depth" here and below is misleading - I believe you mean "offset". The things don't form a two dimensional data structure, so the notion of "depth" is not defined.
</blockquote></details>

Yes, offset is the word I was looking for, updated. I updated the comment a bit too.

---

*[pkg/sql/opt/opt_test.go, line 150 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0kqRquGOqrXXqhpXI4:-L0kxd9Ij49_FlvzGbjE:b-h7qoaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/517aacf7b55788039b83360fc3a615596b136d81/pkg/sql/opt/opt_test.go#L150)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This does not match what you think it matches. (At least it doesn't match what you describe below.)
Can you provide some more examples of what you intend to support? I can help build the re with you.
</blockquote></details>

Fixed (I think?)

---

*[pkg/sql/opt/opt_test.go, line 180 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0kqj1M6I3YLsKGJJ_d:-L0ky9WdQqKu8MZ4toom:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/517aacf7b55788039b83360fc3a615596b136d81/pkg/sql/opt/opt_test.go#L180)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`strings.HasPrefix(strings.TrimLeft(line, " "), "#")`?
Make a function (`stripComments`).
</blockquote></details>

Fixed.

---

*[pkg/sql/opt/opt_test.go, line 361 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0kr6NxRb7aiD7GKLTD:-L0kya6JaYNTPo8Lixxe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/517aacf7b55788039b83360fc3a615596b136d81/pkg/sql/opt/opt_test.go#L361)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

These various functions should be documented.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-:-L0ksVgiUzvqUp7M25zK:bthmf15)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Mostly LGTM. A few questions on the test cases.

---

Reviewed 1 of 4 files at r1, 4 of 4 files at r3.
Review status: 2 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/index_constraints.go, line 166 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0lu4g1_4IsEdtqFADp:-L0lu4g1_4IsEdtqFADq:b-sc430i) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ba0085906130954b7592fc3c8c1e3c074131ed/pkg/sql/opt/index_constraints.go#L166)):*
> ```Go
> 	for i := 0; i < len(a) && i < len(b); i++ {
> 		if cmp := a[i].Compare(c.evalCtx, b[i]); cmp != 0 {
> 			if c.colInfos[offset+i].direction == encoding.Descending {
> ```

If direction were either `-1` or `+1`, you could `return cmp * c.colInfos[offset+i].direction`. 

---

*[pkg/sql/opt/testdata/index-constraints, line 110 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0lti7VhD-TkOpGiDH9:-L0lti7VhD-TkOpGiDHA:b-s08tdv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1235ccc82192629b47a0a326f8fdc6e44521b77c/pkg/sql/opt/testdata/index-constraints#L110)):*
> ```
> (/1.5 - /2)
> 
> # Note the difference here between decimacolumns=(l and int: we
> ```

Something got mangled here.

---

*[pkg/sql/opt/testdata/index-constraints, line 87 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0lvUpLJrqC1QDdpD1P:-L0lvUpLJrqC1QDdpD1Q:b-uzzwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ba0085906130954b7592fc3c8c1e3c074131ed/pkg/sql/opt/testdata/index-constraints#L87)):*
> ```
> @1 >= 1 AND @1 <= 2 AND @2 >= 8 AND @2 <= 9
> ----
> [/2/8 - /1/9]
> ```

Shouldn't the start key be less than the end key?

---

*[pkg/sql/opt/testdata/index-constraints, line 243 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0lvugpVBZo9aWcaIJN:-L0lvugpVBZo9aWcaIJO:bqzejxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ba0085906130954b7592fc3c8c1e3c074131ed/pkg/sql/opt/testdata/index-constraints#L243)):*
> ```
> [/3/4 - /3/4]
> [/2/4 - /2/4]
> [/1/4 - /1/4]
> ```

What is causing the spans to be generated in the opposite order in this case? I'm wondering if that will be problematic as higher levels might like the spans in sorted order.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-:-L0lw72F4w9oFa7di1Da:b-m4yaek)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/testdata/index-constraints, line 87 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0lvUpLJrqC1QDdpD1P:-L0m7XqF9WqOaOzKHR5v:b-uim1he) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ba0085906130954b7592fc3c8c1e3c074131ed/pkg/sql/opt/testdata/index-constraints#L87)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Shouldn't the start key be less than the end key?
</blockquote></details>

It is "less", in that /2 comes before /1 in a descending column.

---

*[pkg/sql/opt/testdata/index-constraints, line 243 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0lvugpVBZo9aWcaIJN:-L0m7cpo9CtYoE2XY5GP:b-p268w6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ba0085906130954b7592fc3c8c1e3c074131ed/pkg/sql/opt/testdata/index-constraints#L243)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

What is causing the spans to be generated in the opposite order in this case? I'm wondering if that will be problematic as higher levels might like the spans in sorted order.
</blockquote></details>

Descending reverses everything for that column; it is "direction-aware" sorted order.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I updated the `preferInclusive` code. It wasn't incorrect but it was unintuitive (we weren't taking into account column directions, but that worked because we were using it on a "reversed" span in the descending case). Now we call it on the correct span, and take directions into account when calling Next or Prev.

---

Review status: 2 of 5 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-:-L0o1WMa4HV7N6B4kRJ0:b-khlrgz)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/testdata/index-constraints, line 87 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0lvUpLJrqC1QDdpD1P:-L0oHdPJjDq2ZowCoJA1:b-3jggk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ba0085906130954b7592fc3c8c1e3c074131ed/pkg/sql/opt/testdata/index-constraints#L87)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It is "less", in that /2 comes before /1 in a descending column.
</blockquote></details>

Ack, though I'm wondering how this plays out with the eventual call to `Scan` or `ReverseScan`. I know `Scan` requires start key to be less than end key. Will it have to check the span and adjust the call?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/testdata/index-constraints, line 87 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0lvUpLJrqC1QDdpD1P:-L0oN9xCjagrOdZ_aLP5:b-ocpp00) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ba0085906130954b7592fc3c8c1e3c074131ed/pkg/sql/opt/testdata/index-constraints#L87)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ack, though I'm wondering how this plays out with the eventual call to `Scan` or `ReverseScan`. I know `Scan` requires start key to be less than end key. Will it have to check the span and adjust the call?
</blockquote></details>

When we encode a LogicalSpan into a KV span, the start key for will be less than the end key (because of the descending encoding).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897)*
<!-- Sent from Reviewable.io -->

knz commented:
The code appears quite well structured overall but I can't judge whether it's correct -- it's much too complex for me to evaluate without a long in-depth review. FWIW each function individually seems to do what its doc indicates it does.

---

Reviewed 1 of 4 files at r3, 1 of 3 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/index_constraints.go, line 166 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0lu4g1_4IsEdtqFADp:-L0oUksBNvxLYLMjgDCR:b-adx75p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ba0085906130954b7592fc3c8c1e3c074131ed/pkg/sql/opt/index_constraints.go#L166)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

If direction were either `-1` or `+1`, you could `return cmp * c.colInfos[offset+i].direction`. 
</blockquote></details>

nit: In an out-of-order processor, it might well be that the current code is faster (cmp/-cmp can be pre-loaded while the conditional is computed). A multiplication introduces a sequential dependency.
I don't really care though.

---

*[pkg/sql/opt/opt_test.go, line 150 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0kqRquGOqrXXqhpXI4:-L0oVHc1v1dlNZoOe8w_:b4fokdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/517aacf7b55788039b83360fc3a615596b136d81/pkg/sql/opt/opt_test.go#L150)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Fixed (I think?)
</blockquote></details>

Your regular expressions accepts things like "a,,,,b" too. 

Try this, this is also more readable:

```go
const thingRe = `[a-zA-Z0-9_-]+`
const commaListRe = thingRe + `(,` + thingRE + `)*`
const maybeEqRe = thingRe + `(=(` + thingRE + `|\(` + commaListRe + `\)))?`
const itemRe = `(` + commaListRe + `|` + maybeEqRe + `)`
const splitDirectivesRE = itemRe + `(\s+` + itemRe + `)*`
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-:-L0oXHVdhbk4dNPWmoDi:b40zh2t)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/index_constraints.go, line 166 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0lu4g1_4IsEdtqFADp:-L0oli7fqzJjxr17adIH:b9cn4l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ba0085906130954b7592fc3c8c1e3c074131ed/pkg/sql/opt/index_constraints.go#L166)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: In an out-of-order processor, it might well be that the current code is faster (cmp/-cmp can be pre-loaded while the conditional is computed). A multiplication introduces a sequential dependency.
I don't really care though.
</blockquote></details>

I left it as is to avoid adding yet another direction type. The common case is "ascending" in which case the `if` is very cheap (branch prediction).

---

*[pkg/sql/opt/opt_test.go, line 150 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897#-L0kqRquGOqrXXqhpXI4:-L0olET6ZAAVbbUpHT0_:be59dlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/517aacf7b55788039b83360fc3a615596b136d81/pkg/sql/opt/opt_test.go#L150)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Your regular expressions accepts things like "a,,,,b" too. 

Try this, this is also more readable:

```go
const thingRe = `[a-zA-Z0-9_-]+`
const commaListRe = thingRe + `(,` + thingRE + `)*`
const maybeEqRe = thingRe + `(=(` + thingRE + `|\(` + commaListRe + `\)))?`
const itemRe = `(` + commaListRe + `|` + maybeEqRe + `)`
const splitDirectivesRE = itemRe + `(\s+` + itemRe + `)*`
```
</blockquote></details>

It doesn't work, it doesn't handle `arg=(val1, <..., may contain spaces>)`. I also don't want to restrict the values to thingRe (for one, they can have spaces). I spent some time trying to fix it but this is a waste of time. Leaving it as it is since it's working. It's ok if it's permissive, the code that actually handles the values can croak later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20897)*
<!-- Sent from Reviewable.io -->
